src/wasm-type.h

```diff
@@ -399,19 +401,31 @@ class HeapType {
   size_t getDepth() const;
 
   // Get the bottom heap type for this heap type's hierarchy.
-  BasicHeapType getBottom() const;
+  BasicHeapType getUnsharedBottom() const;
+  BasicHeapType getBottom() const {
+    return HeapType(getUnsharedBottom()).getSharedBasic(isShared());
+  }
 
   // Get the top heap type for this heap type's hierarchy.
-  BasicHeapType getTop() const;
+  BasicHeapType getUnsharedTop() const;
+  BasicHeapType getTop() const {
+    return HeapType(getUnsharedTop()).getSharedBasic(isShared());
+  }
 
   // Get the recursion group for this non-basic type.
   RecGroup getRecGroup() const;
   size_t getRecGroupIndex() const;
 
   constexpr TypeID getID() const { return id; }
-  constexpr BasicHeapType getBasic() const {
-    assert(isBasic() && "Basic heap type expected");
-    return static_cast<BasicHeapType>(id);
+
+  // Get the shared or unshared version of this basic heap type.
+  constexpr BasicHeapType getSharedBasic(bool shared = true) const {
```

Comment:
How about this?
```cpp
enum Shareability {
  Shared,
  Unshared
};
.. getBasic(Shareability shared = Unshared)
```
Then existing callsites don't need to change, and also it avoids a function `getSharedBasic(..)` which might return a non-Shared result (that seems like it could avoid confusion). There could also be helpers (though maybe they are not needed),
```cpp
getSharedBasic() { return getBasic(Shared); }
getUnsharedBasic() { return getBasic(Unshared); }
```

Reply:
It's important that `getBasic` doesn't exist after this change. If it did exist, users might not consider sharedness when they use it. When the only options are `getUnsharedBasic` and getSharedBasic`, it's much harder to forget to handle sharedness correctly.

Regarding the `Shareability` enum, I don't think it's necessary because there's a clear correspondence with `shared` as true and `unshared` as false. Nobody would think that true mean `unshared` they way they might confuse e.g. nullability. If you disagree and think the enum is worth it, I can add it.

Reply:
Fair enough, I agree it's important to force users to consider sharedness. How about `getBasic(Shareability shared)` then, with no default? I think that's better because of the reason I said before, that it seems slightly confusing for a function `getSharedBasic` to not always return a shared result.

Reply:
The "shared" in `getSharedBasic(bool)` is more like "shared?" in that the parameter says whether it will actually be shared or not. I don't think that's too confusing. I'd prefer `getMaybeSharedBasic(bool)`, but that's way too verbose.

Reply:
I thought on this some more on the weekend and I still think `getSharedBasic(bool)` is not ideal. I get that we could get used to `getSharedBasic(bool)`, but for a new reader, or maybe after not using the API for a time, it isn't immediately obvious if the bool refers to the shared or the basic. That is, the name suggests it is getting something both shared and basic - which does the bool refer to?

It doesn't feel like say `setOpen(bool = true)` where there is just one possibility, so is clear that "open" refers to the bool. That's just a simple setter function, really.

I don't feel strongly between less ambiguous options like `getSharedBasic() / getUnsharedBasic()` or `getBasic(Sharability)`. I just think it would be best if reading the code makes it really obvious what the API does. But maybe there's a better option?

Reply:
I'll try switching to `getGBasic(Shareability)` and see what I think.

Reply:
See the latest commit. I still prefer it the way it was; the APIs using `Shareability` are generally more verbose and it seems like an unnecessary complication over using a bool.

Reply:
I agree this part is less good,
```diff
- typeBuilder[i].setShared(type.isShared());
+ typeBuilder[i].setShareability(type.getShareability());
```
But maybe that could be
```
  typeBuilder[i].setShared(type.getShared());
```
? That is, I think we can call the setter `setShared` and have a getter `getShared`; Shareability can be the enum name, but the getters/setters can be more concise.

Other than that the last commit looks like an improvement to me actually. Even some parts that get a bit longer look like they benefit in clarity,
```diff
- HeapTypeT makeFuncType(bool) { return Ok{}; }
+ HeapTypeT makeFuncType(Shareability) { return Ok{}; }
```
(before, whether the bool was shared or open or some other property is not obvious).

I don't have an opinion on
```diff
- type.getUnsharedBasic()
+ type.getBasic(Unshared)
```
I think we could still use the old form there.

Reply:
Another option is to have (instead, or in addition for convenience) boolean-using getters/setters `setShared(bool), bool isShared()` as those are unambiguous, and use Shareability in ambiguous contexts. We do something similar for signed iirc.

Reply:
Ok, I updated `{get,set}Shareability` to `{get,set}Shared` and for the setters made the default parameter `Shared`. I'm happy with this.